File: scripts/vectors.mjs
// scripts/vectors.mjs — TypeScript side of the NR-4 parity vectors.
// Same inputs as crates/race-phys/examples/vectors.rs. Run:
//   npx tsx scripts/vectors.mjs
//   cargo run --manifest-path crates/race-phys/Cargo.toml --example vectors
// Lines must match (exact f64 parity).
import { stepRacer } from '../games/nitro-rift/physics.js';

const show = (s) => console.log(`${s.prog} ${s.lane} ${s.boost} ${s.slowUntil}`);
const taken = (r) => r.padTaken;

let a = { prog: 0, lane: 1, boost: 100, slowUntil: 0 };
let t = stepRacer(a, { steer: 0, boost: true }, [], [], 0, 1000);
process.stdout.write(`${taken(t)} `); show(a);

let b = { prog: 0, lane: 1, boost: 40, slowUntil: 0 };
t = stepRacer(b, { steer: 0, boost: false }, [], [], 0, 1000);
process.stdout.write(`${taken(t)} `); show(b);

let c = { prog: 0, lane: 1, boost: 10, slowUntil: 0 };
t = stepRacer(c, { steer: 0, boost: false }, [{ at: 3, lane: 1 }], [], 0, 100);
process.stdout.write(`${taken(t)} `); show(c);

let d = { prog: 0, lane: 1, boost: 100, slowUntil: 0 };
t = stepRacer(d, { steer: 0, boost: true }, [], [{ prog: 12, lane: 1 }], 1000, 100);
process.stdout.write(`${taken(t)} `); show(d);
t = stepRacer(d, { steer: 0, boost: true }, [], [], 1100, 1000);
process.stdout.write(`${taken(t)} `); show(d);

let e = { prog: 0, lane: 0, boost: 100, slowUntil: 0 };
t = stepRacer(e, { steer: -1, boost: false }, [], [], 0, 50);
process.stdout.write(`${taken(t)} `); show(e);
t = stepRacer(e, { steer: 1, boost: false }, [], [], 0, 50);
process.stdout.write(`${taken(t)} `); show(e);

File: crates/race-phys/examples/vectors.rs
//! Prints step_racer outputs for the parity vectors (NR-4).
//! Compare against scripts/vectors.mjs (same inputs, TypeScript side).
use race_phys::*;

fn show(s: &RacerState) {
    println!("{} {} {} {}", s.prog, s.lane, s.boost, s.slow_until);
}

fn main() {
    // 1: full boost, 1s
    let mut a = RacerState { prog: 0.0, lane: 1.0, boost: 100.0, slow_until: 0.0 };
    let t = step_racer(&mut a, RaceInput { steer: 0.0, boost: true }, &[], &[], 0.0, 1000.0);
    print!("{} ", t);
    show(&a);
    // 2: coast, 1s
    let mut b = RacerState { prog: 0.0, lane: 1.0, boost: 40.0, slow_until: 0.0 };
    let t = step_racer(&mut b, RaceInput { steer: 0.0, boost: false }, &[], &[], 0.0, 1000.0);
    print!("{} ", t);
    show(&b);
    // 3: pad take at 100ms
    let mut c = RacerState { prog: 0.0, lane: 1.0, boost: 10.0, slow_until: 0.0 };
    let t = step_racer(
        &mut c,
        RaceInput { steer: 0.0, boost: false },
        &[Pad { at: 3.0, lane: 1.0 }],
        &[],
        0.0,
        100.0,
    );
    print!("{} ", t);
    show(&c);
    // 4: bump then slowed second
    let mut d = RacerState { prog: 0.0, lane: 1.0, boost: 100.0, slow_until: 0.0 };
    let t = step_racer(
        &mut d,
        RaceInput { steer: 0.0, boost: true },
        &[],
        &[Other { prog: 12.0, lane: 1.0 }],
        1000.0,
        100.0,
    );
    print!("{} ", t);
    show(&d);
    let t = step_racer(&mut d, RaceInput { steer: 0.0, boost: true }, &[], &[], 1100.0, 1000.0);
    print!("{} ", t);
    show(&d);
    // 5: steer glide + clamp
    let mut e = RacerState { prog: 0.0, lane: 0.0, boost: 100.0, slow_until: 0.0 };
    let t = step_racer(&mut e, RaceInput { steer: -1.0, boost: false }, &[], &[], 0.0, 50.0);
    print!("{} ", t);
    show(&e);
    let t = step_racer(&mut e, RaceInput { steer: 1.0, boost: false }, &[], &[], 0.0, 50.0);
    print!("{} ", t);
    show(&e);
}
